_posts/2018-12-18-blog-a-more-complete-front-end-testing-strategy.md

```diff
@@ -0,0 +1,243 @@
+
+![generic-test](https://raw.githubusercontent.com/unadlib/integration-blog/testing_strategy/_site/assets/generic-test.png)
+
+- E2E
+
+End-to-end testing involves ensuring that the integrated components of an application function as expected. The entire application is tested in a real-world scenario such as communicating with the database, network, hardware and other applications.
+
+- IT
+
+Integration testing is a key software development life cycle strategy. Generally, small software systems are integrated and tested in a single phase, whereas larger systems involve several integration phases to build a complete system, such as integrating modules into low-level subsystems for integration with larger subsystems. Integration testing encompasses all aspects of a software system's performance, functionality and reliability.
+
+
+- UT
+
+Unit testing is a software testing method by which individual units of source code, sets of one or more computer program modules together with associated control data, usage procedures, and operating procedures, are tested to determine whether they are fit for use.
+
+
+
+## What's wrong with the general front-end testing?
+
+When we are evaluating the integrity of a test strategy, we try to validate it with some of the following criteria:
+
+* Pass acceptance criteria
+* Catch bugs as early as possible
+* Faster running speed, especially at the top level of testing
+* Find bugs efficiently
+* Test writing costs and maintenance costs
+* Continuous refactoring risks
+
+A better test strategy should not be as consistent as possible with the above items.
+
+In the general testing strategy mentioned above, the problems posed by the problem, we may not be able to catch bugs as early as possible. Because the E2E test is most likely to run once in a relatively long period of time, rather than every PR or even every code commit, because the e2e run is usually slow and unstable, it is the most expensive automated test to run.
+
+Sometimes we have some integration tests that start a bloated integrated system, maybe it already includes a lot of mock, and you can keep running tests over and over again, but when there are more and more cases of integration testing, we can't even guarantee faster completion within each PR, and in a resource-constrained CI environment, it could be half an hour, or even longer.
+
+As a system becomes more complex, we need a complete testing strategy to tell us what cases failed in these tests, and we were able to catch bugs more efficiently through test reports. Whether it is network instability, back-end server APIs exceptions, front-end domain modules exceptions or UI components exceptions, and so on, we can quickly catch this bugs. Obviously, the general testing strategy can provide a limited amount of lookup bugs help. For example, UT has succeeded, IT has failed, and E2E has failed too. It is difficult for us to analyze clearer information from such test reports.
```

Comment:
instead of 'we were able to catch bugs more efficiently', change it to 'to enable us to catch bugs more efficiently'.
Also, 'we can quickly catch these bugs'. 'This' is singular, and 'these' is plural.
'can provide a limited amount of lookup bugs help' is not good English, maybe 'can provide limited help in locating bugs'.